crates/commitlog/src/buf.rs

```diff
@@ -0,0 +1,114 @@
+use std::{
+    alloc,
+    num::NonZeroUsize,
+    ops::{Deref, DerefMut},
+    ptr,
+};
+
+/// A byte buffer of non-zero size and proper alignment.
+#[derive(Debug)]
+pub struct Aligned {
+    data: ptr::NonNull<u8>,
+    layout: alloc::Layout,
+    size: usize,
+}
+
+impl Aligned {
+    /// Create a new buffer of `size` and `align`.
+    ///
+    /// The parameters are implicitly adjusted if they don't satisfy the
+    /// invariants of [`alloc::Layout`];
+    ///
+    /// * `align` is made a power of two if it isn't already
+    /// * `align` defaults to 512 if it would otherwise overflow
+    /// * the maximum `size` is `isize::MAX - (align - 1)` and will
+    ///   be adjusted if necessary
+    pub fn new(size: NonZeroUsize, align: NonZeroUsize) -> Self {
+        let align = align.get().checked_next_power_of_two().unwrap_or(512);
+        let max_size = isize::MAX as usize - (align - 1);
+        let size = if size.get() > max_size { max_size } else { size.get() };
+
+        // SAFETY: Invariants checked above: `align` is non-zero and a power of
+        // two, `size` rounded up to the nearest multiple of `align` doesn't
+        // overflow isize.
+        let layout = unsafe { alloc::Layout::from_size_align_unchecked(size, align) };
+        // SAFETY: `Layout` has non-zero size.
+        let ptr = unsafe { alloc::alloc(layout) };
+        if ptr.is_null() {
+            alloc::handle_alloc_error(layout);
+        }
+        // SAFETY: We just checked that `ptr` is non-null.
+        let data = unsafe { ptr::NonNull::new_unchecked(ptr) };
+
+        Self { data, layout, size }
+    }
+
+    /// Borrow the buffer as a slice.
+    pub fn as_bytes(&self) -> &[u8] {
+        // SAFETY: The allocation has non-zero size `self.size` and is properly
+        // aligned. `self.size` cannot be modified outside `Self`.
+        unsafe { std::slice::from_raw_parts(self.data.as_ptr(), self.size) }
```

Comment:
This method may create a reference to uninitialized memory, causing UB. It seems like `Aligned` should be a `NonNull<MaybeUninit<u8>>`, and this method should return `&[MaybeUninit<u8>]`.

Reply:
Is it? I mean, I really don’t know what I’m doing, but I thought the size parameter of the layout would mean it gets initialized.

I think the memory should be allocated up front instead of lazily. What’s the right way to do that?

Reply:
The memory is allocated up-front, but not initialized. This is the difference between `alloc` and `alloc_zeroed`. From an abstract machine perspective, the memory returned by `alloc` initially holds the special value `uninit`/`poison`, which is not a valid `u8`, but is a valid `MaybeUninit<u8>`. Constructing an `&u8` or `&mut u8` to `uninit` is UB, so you have to either allocate an initialized buffer with `alloc_zeroed`, or use `&MaybeUninit<u8>` instead.

Reply:
I see, thanks.

I took this from glommio btw, I guess I should maybe let them know.